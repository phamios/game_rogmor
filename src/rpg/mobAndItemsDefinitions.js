export const SEX = {
  male: 'male',
  female: 'female',
};

export const mobDefinitions = [
  {title: "", descript: "elf girl", sex: SEX.female },
  {title: "", descript: "box face", sex: SEX.male },
  {title: "", descript: "chaotic hair", sex: SEX.male },
  {title: "", descript: "bold face", sex: SEX.male },
  {title: "", descript: "fat face", sex: SEX.male },
  {title: "", descript: "charming knight", sex: SEX.male },
  {title: "", descript: "elf merchant", sex: SEX.male },
  {title: "", descript: "cow one", sex: SEX.female },
  {title: "", descript: "elf bard", sex: SEX.male },
  {title: "", descript: "elf girl with dark hair", sex: SEX.female },
  {title: "", descript: "young troll", sex: SEX.male },
  {title: "", descript: "horned inspector", sex: SEX.male },
  {title: "", descript: "citizen", sex: SEX.male },
  {title: "", descript: "blond hair", sex: SEX.female },
  {title: "", descript: "pig one", sex: SEX.male },
  {title: "", descript: "sqared orc", sex: SEX.male },
  {title: "", descript: "bar tender ogre", sex: SEX.male },
  {title: "", descript: "wizard", sex: SEX.male },
  {title: "", descript: "handsome human", sex: SEX.male },
  {title: "", descript: "neighbour", sex: SEX.male },
  {title: "", descript: "long hair", sex: SEX.male },
  {title: "", descript: "six horn", sex: SEX.male },
  {title: "", descript: "desert warrior", sex: SEX.male },
  {title: "", descript: "sad fighter", sex: SEX.male },
  {title: "", descript: "trader", sex: SEX.male },
  {title: "", descript: "paladin", sex: SEX.male },
  {title: "", descript: "knight girl", sex: SEX.female },
  {title: "", descript: "adventurer", sex: SEX.male },
  {title: "", descript: "psycho", sex: SEX.male },
];

export const itemDefinitions = [
  {title: "", descript: "saw knife"},
  {title: "", descript: "short sword"},
  {title: "", descript: "halberd"},
  {title: "", descript: "bow"},
  {title: "", descript: "iron shield"},
  {title: "", descript: "rounded shield"},
  {title: "", descript: "sword"},
  {title: "", descript: "dagger"},
  {title: "", descript: "spiked jerk"},
  {title: "", descript: "staff"},
  {title: "", descript: "bishop staff"},
  {title: "", descript: "axe"},
  {title: "", descript: "ring"},
  {title: "", descript: "bold blade"},
  {title: "", descript: "rapier"},
  {title: "", descript: "hammer"},
  {title: "", descript: "recoil bow"},
  {title: "", descript: "xiphos"},
  {title: "", descript: "harpoon"},
  {title: "", descript: "spear"},
  {title: "", descript: "great shield"},
  {title: "", descript: "knife"},
  {title: "", descript: "mallet"},
  {title: "", descript: "long sword"},
  {title: "", descript: "dual axe"},
  {title: "", descript: "sabre"},
  {title: "", descript: "knight sword"},
  {title: "", descript: "crossbow"},
  {title: "", descript: "claw"},
  {title: "", descript: "broad sword"},
  {title: "", descript: "barbarian axe"},
  {title: "", descript: "saw dagger"},
  {title: "", descript: "cup"},
  {title: "", descript: "material"},
  {title: "", descript: "chest"},
  {title: "", descript: "bag"},
  {title: "", descript: "coin"},
  {title: "", descript: "anvil"},
  {title: "", descript: "box"},
  {title: "", descript: "potion"},
  {title: "", descript: "stone"},
  {title: "", descript: "chicken"},
  {title: "", descript: "goblet"},
  {title: "", descript: "flag"},
  {title: "", descript: "point of interest"},
];