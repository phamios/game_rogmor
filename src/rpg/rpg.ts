export const leveling = (level:number, mod:number = 0.5, pow:number = 4, madd:number = 2) => Math.round(
  mod * Math.pow(level, pow) + (madd * level)
);

export const rnd = (dice:number) => dice * Math.random() | 0;

export const dice = (roll:number) => rnd(roll - 1) + 1;

// export const improved = ( attribute = 1, percent = 0.2 ) => (attribute + (Math.random() * percent * attribute)) | 0;
export const improved = ( attribute:number = 1 ) => attribute * 1.15 | 0;

export const shuffle = () => Math.random() > 0.5 ? 1 : -1;

export const pickOne = (arr:any[]) => arr[rnd(arr.length)];

export const selectRandomPick = (element:any, index:number, arr:any[], pickIndex = Math.random() * arr.length | 0) => index === pickIndex;

export const uid = () => Math.random().toString(32).slice(-8);

export const amount = (n:number) => Array(n).fill(0).map((_, index) => index);
