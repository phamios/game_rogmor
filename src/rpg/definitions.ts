import { Mob } from "./profession";

export interface Effect {
  id: string;
  type: string;
}
export interface Act {
  id: string;
  type: string;
  cooldown: number;
  tick: number;
  script: string;
}

export interface ActInstance {
  act: Act;
  actor: Mob;
  targetList: Mob[];
  timestamp?: number;
}

export interface Encounter {
  mobList: Mob[];
  actList: ActInstance[];
  targetList: Mob[];
  showTime: Effect[];
}

export type OrderOfSeed = [Mob, number];

export enum Outcome { ENDED };
export type OutcomeList = Outcome[];
