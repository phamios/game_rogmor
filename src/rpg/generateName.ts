import { rnd } from "./rpg";

export const ROGMOR_NAME_SOURCE = 'proedmorainowerialaenderomanulnivosudanterill';

export const generateName = (source = ROGMOR_NAME_SOURCE) => {
  const name:string = Array.from(
      {length:2 + rnd(3)}, 
      (s:number = rnd(source.length)) => source.slice(s, s + 2 + rnd(3))
    )
    .join('')
  ;
  const [capital, ...rest] = name.split('');
  return [capital.toUpperCase(),...rest].join('');
}
