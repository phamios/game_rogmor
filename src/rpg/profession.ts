import { rnd, leveling, uid, pickOne } from './rpg';

export interface Ability { 
  title:string;
  body: number;
  reflex: number;
  soul: number;
  aura: number;
  stamina: number;
  focus: number;
  morale: number; 
}

export interface Condition {
  staminaState: number;
  focusState: number;
  moraleState: number;
  isOut?: boolean;
}

export enum Team { PLAYER, GOOD, BAD, UGLY };

export interface MobTraits {  
  level: number;
  professionType: ProfessionKey;
  ability: Ability;
  condition: Condition;
}
export interface MobPerson {  
  name: string;
  avatar: number;
  coord: number;
  uid: string;
  team: Team;
}

export interface Mob extends MobTraits, MobPerson {}

export type ProfessionKey =
  'fighter' |'priest' |'paladin' |'knight' |'assasin' |'rogue' |'mage' |'captain' |'noble' |
  'headHunter' |'witch' |'valkur' |'guard' |'gladiator' |'barbarian' |'wizard' |'sage' |
  'samurai' |'ninja' |'warlord' |'bishop' |'tracker' |'dementor' |'necromanta' |'summoner' |
  'shaman' |'thief' |'rouge' |'cavalier' |'pirate' |'inquisitor' |'icelander' |'heretic' |
  'nomad' |'bard' |'warlock' |'monk' |'druid' |'illusionist' |'wardancer' |'runeSmith' |
  'merchant' |'explorer' |'templar' |'spy' |'nun' |'mercenary' |'duelist' |'beastmaster' |
  'sorcerer' |'archer' |'psionic' |'alchemist' |'enginer' |'antiPaladin' |'chaosKnight' |
  'chief' |'maruder' |'rich' |'shapechanger' 

export type ProfessionLookup = Record<ProfessionKey, Ability>; 

export const professionLookup:ProfessionLookup = {
    fighter:      { title:'Fighter',      body: 7, reflex: 4, soul: 2, aura: 4, stamina: 55, focus: 45, morale: 40 },
    priest:       { title:'Priest',       body: 3, reflex: 3, soul: 7, aura: 8, stamina: 35, focus: 65, morale: 40 },
    paladin:      { title:'Paladin',      body: 6, reflex: 3, soul: 6, aura: 6, stamina: 50, focus: 50, morale: 40 },
    knight:       { title:'Knight',       body: 7, reflex: 3, soul: 3, aura: 5, stamina: 60, focus: 50, morale: 40 },
    assasin:      { title:'Assasin',      body: 5, reflex: 7, soul: 1, aura: 3, stamina: 45, focus: 55, morale: 40 },
    rogue:        { title:'Rogue',        body: 4, reflex: 6, soul: 3, aura: 5, stamina: 55, focus: 40, morale: 40 },
    mage:         { title:'Mage',         body: 3, reflex: 4, soul: 6, aura: 2, stamina: 25, focus: 70, morale: 40 },
    captain:      { title:'Captain',      body: 5, reflex: 5, soul: 5, aura: 6, stamina: 55, focus: 55, morale: 40 },
    noble:        { title:'Noble',        body: 4, reflex: 5, soul: 6, aura: 7, stamina: 45, focus: 60, morale: 40 },
    headHunter:   { title:'Head hunter',  body: 5, reflex: 7, soul: 3, aura: 4, stamina: 60, focus: 45, morale: 40 },
    witch:        { title:'Witch',        body: 2, reflex: 5, soul: 6, aura: 2, stamina: 25, focus: 65, morale: 40 },
    valkur:       { title:'Valkur',       body: 5, reflex: 6, soul: 4, aura: 2, stamina: 60, focus: 60, morale: 40 },
    guard:        { title:'Guard',        body: 6, reflex: 5, soul: 2, aura: 4, stamina: 65, focus: 40, morale: 40 },
    gladiator:    { title:'Gladiator',    body: 7, reflex: 5, soul: 2, aura: 2, stamina: 75, focus: 25, morale: 40 },
    barbarian:    { title:'Barbarian',    body: 8, reflex: 6, soul: 2, aura: 2, stamina: 80, focus: 35, morale: 40 },
    wizard:       { title:'Wizard',       body: 1, reflex: 2, soul: 8, aura: 2, stamina: 25, focus: 80, morale: 40 },
    sage:         { title:'Sage',         body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    samurai:      { title:'Samurai',      body: 6, reflex: 7, soul: 4, aura: 5, stamina: 70, focus: 55, morale: 40 },
    ninja:        { title:'Ninja',        body: 4, reflex: 8, soul: 3, aura: 3, stamina: 55, focus: 45, morale: 40 },
    warlord:      { title:'Warlord',      body: 4, reflex: 5, soul: 6, aura: 8, stamina: 45, focus: 60, morale: 40 },
    bishop:       { title:'Bishop',       body: 4, reflex: 5, soul: 6, aura: 8, stamina: 45, focus: 60, morale: 40 },
    tracker:      { title:'Tracker',      body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    dementor:     { title:'Dementor',     body: 6, reflex: 4, soul: 1, aura: 1, stamina: 75, focus: 70, morale: 40 },
    necromanta:   { title:'Necromanta',   body: 2, reflex: 5, soul: 6, aura: 1, stamina: 45, focus: 60, morale: 40 },
    summoner:     { title:'Summoner',     body: 3, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    shaman:       { title:'Shaman',       body: 4, reflex: 5, soul: 6, aura: 6, stamina: 45, focus: 60, morale: 40 },
    thief:        { title:'Thief',        body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    rouge:        { title:'Rouge',        body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    cavalier:     { title:'Cavalier',     body: 6, reflex: 5, soul: 6, aura: 4, stamina: 45, focus: 60, morale: 40 },
    pirate:       { title:'Pirate',       body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    inquisitor:   { title:'Inquisitor',   body: 4, reflex: 5, soul: 6, aura: 4, stamina: 45, focus: 60, morale: 40 },
    icelander:    { title:'Icelander',    body: 7, reflex: 6, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    heretic:      { title:'Heretic',      body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    nomad:        { title:'Nomad',        body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    bard:         { title:'Bard',         body: 4, reflex: 5, soul: 6, aura: 6, stamina: 45, focus: 60, morale: 40 },
    warlock:      { title:'Warlock',      body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    monk:         { title:'Monk',         body: 6, reflex: 6, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    druid:        { title:'Druid',        body: 4, reflex: 5, soul: 6, aura: 4, stamina: 45, focus: 60, morale: 40 },
    illusionist:  { title:'Illusionist',  body: 3, reflex: 6, soul: 5, aura: 5, stamina: 35, focus: 50, morale: 40 },
    wardancer:    { title:'Wardancer',    body: 5, reflex: 7, soul: 6, aura: 2, stamina: 60, focus: 40, morale: 40 },
    runeSmith:    { title:'Rune smith',   body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    merchant:     { title:'Merchant',     body: 3, reflex: 3, soul: 6, aura: 6, stamina: 55, focus: 65, morale: 40 },
    explorer:     { title:'Explorer',     body: 4, reflex: 5, soul: 6, aura: 7, stamina: 45, focus: 60, morale: 40 },
    templar:      { title:'Templar',      body: 6, reflex: 3, soul: 6, aura: 5, stamina: 60, focus: 55, morale: 40 },
    spy:          { title:'Spy',          body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    nun:          { title:'Nun',          body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    mercenary:    { title:'Mercenary',    body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    duelist:      { title:'Duelist',      body: 3, reflex: 8, soul: 5, aura: 5, stamina: 50, focus: 60, morale: 40 },
    beastmaster:  { title:'Beastmaster',  body: 6, reflex: 5, soul: 7, aura: 1, stamina: 45, focus: 70, morale: 40 },
    sorcerer:     { title:'Sorcerer',     body: 3, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    archer:       { title:'Archer',       body: 4, reflex: 7, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    psionic:      { title:'Psionic',      body: 4, reflex: 4, soul: 7, aura: 3, stamina: 45, focus: 60, morale: 40 },
    alchemist:    { title:'Alchemist',    body: 3, reflex: 5, soul: 7, aura: 2, stamina: 35, focus: 70, morale: 40 },
    enginer:      { title:'Enginer',      body: 3, reflex: 5, soul: 4, aura: 4, stamina: 35, focus: 65, morale: 40 },
    antiPaladin:  { title:'Anti Paladin', body: 7, reflex: 3, soul: 1, aura: 5, stamina: 60, focus: 60, morale: 10 },
    chaosKnight:  { title:'Chaos Knight', body: 7, reflex: 4, soul: 6, aura: 1, stamina: 50, focus: 60, morale: 10 },
    chief:        { title:'Chief',        body: 5, reflex: 5, soul: 4, aura: 7, stamina: 55, focus: 60, morale: 40 },
    maruder:      { title:'Maruder',      body: 7, reflex: 4, soul: 2, aura: 4, stamina: 60, focus: 40, morale: 40 },
    rich:         { title:'Rich',         body: 3, reflex: 4, soul: 3, aura: 8, stamina: 30, focus: 70, morale: 30 },
    shapechanger: { title:'Shapechanger', body: 6, reflex: 4, soul: 4, aura: 2, stamina: 55, focus: 45, morale: 30 },
}

// https://en.wikipedia.org/wiki/Character_class_(Dungeons_%26_Dragons)
// https://wfrp1e.fandom.com/wiki/List_Of_Advanced_Careers

export const professionList = Object.values(professionLookup);
export const professionKeyList = Object.keys(professionLookup) as ProfessionKey[];

export const traitsFactory = (level:number = 1,  professionType:ProfessionKey) => {

  const ability:Ability = professionLookup?.[professionType] ?? pickOne(professionList);

  const stamina = leveling(level, ability.stamina  / 10, 2);
  const focus    = leveling(level, ability.focus     / 10, 2);
  const morale  = leveling(level, ability.morale   / 10, 2);

  const mobTraits:MobTraits = {
    level,
    professionType,
    ability: {
      title        : ability.title,
      body         : leveling(level, ability.body     / 10, 2),
      reflex     : leveling(level, ability.reflex / 10, 2),
      soul         : leveling(level, ability.soul     / 10, 2),
      aura      : leveling(level, ability.aura  / 10, 2),
      stamina,
      focus,
      morale,
    },
    condition: {
      staminaState : stamina,
      focusState    : focus,
      moraleState  : morale,
    }
  };

  return mobTraits;
}

export const mobFactory = (name, avatar, coord, uid, team, traits:MobTraits):Mob => ({
  name,
  avatar,
  coord,
  uid,
  team,
  ...traits
});

export const increaseLevel = (amount:number = 1) => ({
  level, 
  professionType, 
  ability, 
  condition, 
  ...rest
}:Mob):Mob => ({
  ...traitsFactory(level + amount, professionType),
  ...rest
});

export const oneLevelUp = increaseLevel(1);