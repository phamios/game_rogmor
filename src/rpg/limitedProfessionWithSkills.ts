import { ProfessionKey } from "./profession";

export const limitedProfessionWithSkills:ProfessionKey[] = [
  'chaosKnight',
  'bishop',
  'icelander',
  'ninja',
  'samurai',
  'merchant',
];

export const skillForProf:Partial<Record<ProfessionKey, string[]>> = {
  'chaosKnight': ['instant target-rnd hit-soul power-1; instant target-rnd hit-body power-1; instant target-rnd hit-aura power-1','fill-3 target hit-body power-4','fill-4 target-all hit-soul power-2'],
  'bishop': ['instant target-ally heal; instant target-rnd hit-body', 'instant target hit-soul power-2','fill-3 tsa heal-2','fill-4 target-all-ally heal-3'],
  'icelander': ['instant target-all hit-body','fill-3 target-all hit-body power-2','fill-4 target-rnd hit-body power-4'],
  'ninja': ['instant target hit-body power-[2.4]','fill-4 target-all hit-body power-2','fill-2 target hit-body power-4'],
  'samurai': ['instant target hit-body power-2','fill-2 target hit-body power-4','fill-4 target-all hit-body power-2'],
  'merchant': ['instant target hit-aura; instant target-rnd hit-soul power-2','fill-2 target hit-aura power-2','fill-4 target-all hit-aura'],
};