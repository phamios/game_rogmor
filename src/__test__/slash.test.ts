import { Mob, mobFactory, Team, traitsFactory, ProfessionKey } from "../rpg/profession";
import { 
  actionOrder, Target, HitType, SlashObject, SlashParser, slashParse,
  Doit, FlowAction, getSkillResult, skillReducer,
 } from "../rpg/slash";

const makeMob = (lvl:number, prof:ProfessionKey, team:Team = Team.GOOD) => mobFactory(
  `${prof} level:${lvl}`, 2, 1, `id:${prof}-${lvl}`, team, traitsFactory(lvl, prof)
);

const skillForProf:Partial<Record<ProfessionKey, string[]>> = {
  'assasin': ['instant target hit-body','fill-3 hit-body power-4','fill-4 target-all hit-soul power-2'],
  'bishop': ['instant target hit-soul power-2','f-3 tsa heal-2','fill-4 target-all-ally bless-body power-2'],
  'icelander': ['instant target-all hit-body','fill-3 target-all hit-body','fill-4 target-rnd power-4'],
  'ninja': ['instant target hit-body power-[2.4]','fill-4 target-all hit-body power-2','fill-2 target hit-body stun-2 power-4'],
  'samurai': ['instant target hit-body power-2','fill-2 target hit-body power-4','fill-4 target-all hit-body power-2'],
  'merchant': ['instant target hit-body weak','fill-2 hit-aura power-2','fill-4 target-all bribe-2'],
};

const getSkillObject = (m:Mob):Partial<SlashObject>[] => skillForProf[m.professionType].map(slashParse);

const testTeams = [
  [5, 'icelander', Team.BAD],
  [6, 'bishop', Team.BAD],
  [5, 'assasin', Team.BAD],
  [4, 'ninja', Team.GOOD],
  [7, 'samurai', Team.GOOD],
  [4, 'merchant', Team.GOOD],
]

const encounter = {
  mobList: testTeams.map(([lvl, type, team]:[number, ProfessionKey, Team]) => makeMob(lvl, type, team))
}

const slashPreParse = (slashSource:string) => slashSource.split(' ')
  .map(command => command.split('-'))
  .map(([c,m]) => m ? {c,m} : {c}) // c: command, m: modification
;

export function * combatFlowSaga(mobList:Mob[]):any {
  while (true) {
    const order = actionOrder(mobList);
    yield order;
    while (order.length) {
      const [actor]:[Mob] = order.shift();
      yield actor.uid;
      const skillList = getSkillObject(actor);
      yield skillList;
      const [A1] = skillList;
      const [aiTargetting, skillResult] = getSkillResult(actor, A1, mobList);
      mobList = skillReducer(mobList, skillResult);
      yield aiTargetting;
      yield skillResult;
    }
    // yield mobList.map((mob:Mob) => [mob.uid,mob.condition])
  }
}

test ('samurai lvl 5', () => {
  expect (makeMob(5, 'samurai', Team.GOOD)).toMatchSnapshot();
});

test ('encounter setup', () => {
  expect (encounter).toMatchSnapshot();
});

test ('simple strike test', () => {  
  expect (
    slashPreParse('instant target hit-body power-2')
  ).toStrictEqual([
    {c:'instant'},
    {c:'target'},
    {c:'hit', m:'body'},
    {c:'power', m:'2'},
  ]);
});

test.skip ('simple strike test', () => {  
  expect (
    slashParse('i target hb power-[2.4] extra-[teleport]')
  ).toStrictEqual({
    fill:0,
    select: Target.SELECTED_ENEMY,
    type: HitType.BODY,
    doit: Doit.HIT,
    mul: 2.4,
    extra: 'teleport',
  });
});

test ('search unexsist skill', () => {

  const skillList = encounter.mobList.map((hero:Mob) => skillForProf[hero.professionType]
    .map(slashParse)
  ).flat()
  
  expect (
    skillList.filter((skill:any) => skill._ERROR_)
  ).toStrictEqual(
    []
  );
});

describe ('First full combat round', () => {

  const mobList = testTeams.map(([lvl, type, team]:[number, ProfessionKey, Team]) => makeMob(lvl, type, team));
  const flow = combatFlowSaga(mobList)
  const nextRound = (generator) => generator.next().value; 
  const skippRound = (amount:number) => Array(amount).fill(1).forEach(() => nextRound(flow))

  test('initial order', () => {
  
    expect ( 
      nextRound(flow).map(([m]:[Mob])=>[m.uid, m.condition])
    ).toStrictEqual(
      [
        ['id:samurai-7', {staminaState: 357, focusState: 284, moraleState: 210}],
        ['id:bishop-6', {staminaState: 174, focusState: 228, moraleState: 156}],
        ['id:icelander-5', {staminaState: 123, focusState: 160, moraleState: 110}],
        ['id:assasin-5', {staminaState: 123, focusState: 148, moraleState: 110}],
        ['id:ninja-4', {staminaState: 96, focusState: 80, moraleState: 72}],
        ['id:merchant-4', {staminaState: 96, focusState: 112, moraleState: 72}],
      ]
    );
  })

  test ('first is the samurai', () => {
    expect(nextRound(flow)).toStrictEqual('id:samurai-7');
  });

  test ('samurai skill object list is', () => {
    expect(nextRound(flow)).toMatchSnapshot() 
  });

  test ('samurai select one enemy for A1', () => {
    expect(nextRound(flow)).toStrictEqual(
      {
        who: 'id:samurai-7',
        doit: Doit.TARGET,
        select: Target.SELECTED_ENEMY,
        target: ['id:assasin-5'],
      } as FlowAction
    )
  });

  test ('samurai hit weakest enemy', () => {
    expect(nextRound(flow)).toStrictEqual(
      {
        who: 'id:samurai-7',
        doit: Doit.HIT,
        type: HitType.BODY,
        target: ['id:assasin-5'],
        amount: [['id:assasin-5', -49, 74, {staminaState: 74, moraleState: 110, focusState: 148}]],
      } as FlowAction
    )
  });

  test ('samurai A2 and A3 fiiled up bye one', () => {});

  test ('next one is the bishop', () => {
    expect (nextRound(flow)).toStrictEqual('id:bishop-6');
    nextRound(flow)
  });

  test ('bishop select one enemy for A1', () => {
    expect(nextRound(flow)).toStrictEqual(
      {
        who: 'id:bishop-6',
        doit: Doit.TARGET,
        select: Target.SELECTED_ENEMY,
        target: ['id:ninja-4'],
      } as FlowAction
    )
  });

  test ('bishop hit weakest enemy', () => {
    expect(nextRound(flow)).toStrictEqual(
      {
        who: 'id:bishop-6',
        doit: Doit.HIT,
        type: HitType.SOUL,
        target: ['id:ninja-4'],
        amount: [['id:ninja-4', -39, 41, {staminaState: 96, moraleState: 72, focusState: 41}]],
      } as FlowAction
    )
  });

  test ('next one is the icelander', () => {
    expect (nextRound(flow)).toStrictEqual('id:icelander-5');
    nextRound(flow)
  });

  test ('icelander select one enemy for A1', () => {
    expect(nextRound(flow)).toStrictEqual(
      {
        who: 'id:icelander-5',
        doit: Doit.TARGET,
        select: Target.ALL_ENEMY,
        target: [
          'id:ninja-4',  
          'id:samurai-7',
          'id:merchant-4',
        ],
      } as FlowAction
    )
  });

  test ('icelander hit all enemy', () => {
    expect(nextRound(flow)).toStrictEqual(
      {
        who: 'id:icelander-5',
        doit: Doit.HIT,
        type: HitType.BODY,
        target: [
          'id:ninja-4',
          'id:samurai-7',
          'id:merchant-4',
        ],
        amount: [
          ['id:ninja-4', -16, 80, {staminaState: 80, moraleState: 72, focusState: 41}],
          ['id:samurai-7', -16, 341, {staminaState: 341, moraleState: 210, focusState: 284}], 
          ['id:merchant-4', -16, 80, {staminaState: 80, moraleState: 72, focusState: 112}],
        ],
      } as FlowAction
    )
  });
  
  test ('end of the round', () => {
    skippRound(12);
    expect (
      nextRound(flow).map(([m]:[Mob])=>[m.uid, m.condition])
    ).toStrictEqual(
      [
        ['id:samurai-7', {staminaState: 341, focusState: 284, moraleState: 210}],
        ['id:bishop-6', {staminaState: 174, focusState: 228, moraleState: 156}],
        ['id:icelander-5', {staminaState: 123, focusState: 160, moraleState: 110}],
        ['id:assasin-5', {staminaState: 49, focusState: 148, moraleState: 110}],
        ['id:ninja-4', {staminaState: 80, focusState: 41, moraleState: 72}],
        ['id:merchant-4', {staminaState: 67, focusState: 112, moraleState: 72}],
      ]
    );
  });

  test ('assasin knocked out of fight', () => {    
    skippRound(3)
    expect (
      nextRound(flow)
    ).toStrictEqual(
      {
        who: 'id:samurai-7',
        doit: Doit.HIT,
        type: HitType.BODY,
        target: ['id:assasin-5'],
        amount: [['id:assasin-5', -49, 0, {staminaState: 0, moraleState: 110, focusState: 148, isOut: true}]],
      } as FlowAction

    );
    skippRound(2);
  });

  test ('end of the second', () => {
    skippRound(6 * 3)
    expect (
      nextRound(flow).map(([m]:[Mob])=>[m.uid, m.condition])
    ).toStrictEqual(
      [
        ['id:samurai-7', {staminaState: 325, focusState: 284, moraleState: 210}],
        ['id:bishop-6', {staminaState: 174, focusState: 228, moraleState: 156}],
        ['id:icelander-5', {staminaState: 98, focusState: 160, moraleState: 110}],
        // ['id:assasin-5', {staminaState: 0, focusState: 148, moraleState: 110, isOut: true}],
        ['id:merchant-4', {staminaState: 38, focusState: 112, moraleState: 72}],
        ['id:ninja-4', {staminaState: 64, focusState: 2, moraleState: 72}],
      ]
    );
  });

  test ('assasin already out of combat so the last is merchant', () => {
    skippRound(12)
    expect (
      nextRound(flow) //.map(([m]:[Mob])=>[m.uid, m.condition])
    ).toStrictEqual(
      "id:merchant-4"
    );
  });

});


