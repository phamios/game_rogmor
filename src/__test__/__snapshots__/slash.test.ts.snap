// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First full combat round samurai skill object list is 1`] = `
Array [
  Object {
    "doit": 2,
    "fill": 0,
    "mul": 2,
    "select": 6,
    "type": 1,
  },
  Object {
    "doit": 2,
    "fill": 2,
    "mul": 4,
    "select": 6,
    "type": 1,
  },
  Object {
    "doit": 2,
    "fill": 4,
    "mul": 2,
    "select": 2,
    "type": 1,
  },
]
`;

exports[`encounter setup 1`] = `
Object {
  "mobList": Array [
    Object {
      "ability": Object {
        "aura": 18,
        "body": 28,
        "focus": 160,
        "morale": 110,
        "reflex": 25,
        "soul": 25,
        "stamina": 123,
        "title": "Icelander",
      },
      "avatar": 2,
      "condition": Object {
        "focusState": 160,
        "moraleState": 110,
        "staminaState": 123,
      },
      "coord": 1,
      "level": 5,
      "name": "icelander level:5",
      "professionType": "icelander",
      "team": 2,
      "uid": "id:icelander-5",
    },
    Object {
      "ability": Object {
        "aura": 41,
        "body": 26,
        "focus": 228,
        "morale": 156,
        "reflex": 30,
        "soul": 34,
        "stamina": 174,
        "title": "Bishop",
      },
      "avatar": 2,
      "condition": Object {
        "focusState": 228,
        "moraleState": 156,
        "staminaState": 174,
      },
      "coord": 1,
      "level": 6,
      "name": "bishop level:6",
      "professionType": "bishop",
      "team": 2,
      "uid": "id:bishop-6",
    },
    Object {
      "ability": Object {
        "aura": 18,
        "body": 23,
        "focus": 148,
        "morale": 110,
        "reflex": 28,
        "soul": 13,
        "stamina": 123,
        "title": "Assasin",
      },
      "avatar": 2,
      "condition": Object {
        "focusState": 148,
        "moraleState": 110,
        "staminaState": 123,
      },
      "coord": 1,
      "level": 5,
      "name": "assasin level:5",
      "professionType": "assasin",
      "team": 2,
      "uid": "id:assasin-5",
    },
    Object {
      "ability": Object {
        "aura": 13,
        "body": 14,
        "focus": 80,
        "morale": 72,
        "reflex": 21,
        "soul": 13,
        "stamina": 96,
        "title": "Ninja",
      },
      "avatar": 2,
      "condition": Object {
        "focusState": 80,
        "moraleState": 72,
        "staminaState": 96,
      },
      "coord": 1,
      "level": 4,
      "name": "ninja level:4",
      "professionType": "ninja",
      "team": 1,
      "uid": "id:ninja-4",
    },
    Object {
      "ability": Object {
        "aura": 39,
        "body": 43,
        "focus": 284,
        "morale": 210,
        "reflex": 48,
        "soul": 34,
        "stamina": 357,
        "title": "Samurai",
      },
      "avatar": 2,
      "condition": Object {
        "focusState": 284,
        "moraleState": 210,
        "staminaState": 357,
      },
      "coord": 1,
      "level": 7,
      "name": "samurai level:7",
      "professionType": "samurai",
      "team": 1,
      "uid": "id:samurai-7",
    },
    Object {
      "ability": Object {
        "aura": 18,
        "body": 13,
        "focus": 112,
        "morale": 72,
        "reflex": 13,
        "soul": 18,
        "stamina": 96,
        "title": "Merchant",
      },
      "avatar": 2,
      "condition": Object {
        "focusState": 112,
        "moraleState": 72,
        "staminaState": 96,
      },
      "coord": 1,
      "level": 4,
      "name": "merchant level:4",
      "professionType": "merchant",
      "team": 1,
      "uid": "id:merchant-4",
    },
  ],
}
`;

exports[`samurai lvl 5 1`] = `
Object {
  "ability": Object {
    "aura": 23,
    "body": 25,
    "focus": 148,
    "morale": 110,
    "reflex": 28,
    "soul": 20,
    "stamina": 185,
    "title": "Samurai",
  },
  "avatar": 2,
  "condition": Object {
    "focusState": 148,
    "moraleState": 110,
    "staminaState": 185,
  },
  "coord": 1,
  "level": 5,
  "name": "samurai level:5",
  "professionType": "samurai",
  "team": 1,
  "uid": "id:samurai-5",
}
`;
